src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/DoNotGuardSetAddOrRemoveByContains.cs

```diff
@@ -0,0 +1,200 @@
+            {
+                return;
+            }
+
+            if (!AreInvocationsOnSameInstance(containsInvocation, addOrRemoveInvocation))
+            {
+                return;
+            }
+
+            using var locations = ArrayBuilder<Location>.GetInstance(2);
+            locations.Add(conditional.Syntax.GetLocation());
+            locations.Add(addOrRemoveInvocation.Syntax.Parent!.GetLocation());
+
+            context.ReportDiagnostic(containsInvocation.CreateDiagnostic(Rule, additionalLocations: locations.ToImmutable(), null));
+        }
+
+        private static bool TryGetRequiredMethods(
+            Compilation compilation,
+            [NotNullWhen(true)] out IMethodSymbol? containsMethod,
+            [NotNullWhen(true)] out IMethodSymbol? addMethod,
+            [NotNullWhen(true)] out IMethodSymbol? removeMethod)
+        {
+            var iSetType = WellKnownTypeProvider.GetOrCreate(compilation).GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemCollectionsGenericISet1);
+            var iCollectionType = WellKnownTypeProvider.GetOrCreate(compilation).GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemCollectionsGenericICollection1);
+
+            if (iSetType is null || iCollectionType is null)
+            {
+                containsMethod = null;
+                addMethod = null;
+                removeMethod = null;
+
+                return false;
+            }
+
+            addMethod = iSetType.GetMembers(Add).OfType<IMethodSymbol>().FirstOrDefault();
+            containsMethod = iCollectionType.GetMembers(Contains).OfType<IMethodSymbol>().FirstOrDefault();
+            removeMethod = iCollectionType.GetMembers(Remove).OfType<IMethodSymbol>().FirstOrDefault();
+
+            return containsMethod is not null && addMethod is not null && removeMethod is not null;
+        }
+
+        private static bool TryExtractContainsInvocation(
+            IOperation condition,
+            IMethodSymbol containsMethod,
+            [NotNullWhen(true)] out IInvocationOperation? containsInvocation,
+            out bool containsNegated)
+        {
+            containsNegated = false;
+            containsInvocation = null;
+
+            switch (condition.WalkDownParentheses())
+            {
+                case IInvocationOperation invocation:
+                    containsInvocation = invocation;
+                    break;
+                case IUnaryOperation unaryOperation when unaryOperation.OperatorKind == UnaryOperatorKind.Not && unaryOperation.Operand is IInvocationOperation operand:
+                    containsNegated = true;
+                    containsInvocation = operand;
+                    break;
+                default:
+                    return false;
+            }
+
+            return DoesSignatureMatch(containsInvocation.TargetMethod, containsMethod);
```

Comment:
Why not compare the symbols directly?

Reply:
The `containsMethod` comes from `ICollection<T>` whereas `containsInvocation.TargetMethod` comes from the concrete implementation (e.g. `HashSet<T>`).
So comparing the symbols directly would lead to inequality, as `System.Collections.Generic.ICollection<T>.Contains(T)` != `System.Collections.Generic.HashSet<string>.Contains(string)`.

That is why I resorted to compare method signatures instead.

But I think a better approach would be to check if the target method implements the interface method, something like this:
```C#
private static bool DoesImplementInterfaceMethod(IMethodSymbol method, IMethodSymbol interfaceMethod)
{
    return method.IsImplementationOfInterfaceMethod(method.Parameters[0].Type, interfaceMethod.ContainingType, interfaceMethod.Name);
}
```

`PreferDictionaryTryMethodsOverContainsKeyGuardAnalyzer` also uses a signature check which could be replaced by the above method:

https://github.com/dotnet/roslyn-analyzers/blob/4592065e23f7c84047b4e798ec41efd9876c41bc/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferDictionaryTryMethodsOverContainsKeyGuardAnalyzer.cs#L263-L268